File: typescript/src/pricing.ts
import type { Provider } from './types.js';

export interface ModelPricing {
  provider: Provider;
  inputPerMillion: number;
  outputPerMillion: number;
  /** Cache-READ input rate per million (discounted repeated context). */
  cachedInputPerMillion: number;
  /** Cache-WRITE rate per million. Anthropic only (write premium); omitted for
   *  OpenAI/Gemini which have no per-request write charge. */
  cacheWritePerMillion?: number;
}

// KEEP IN SYNC with:
//   - weckr-api/lib/caps.ts              (server, authoritative)
//   - weckr-python/weckr/pricing.py      (Python SDK)
// If you add a model here, add it there too.
//
// Cached rates verified against official provider pricing on 2026-07-18:
//   OpenAI gpt-4o + o-series: cache read = 0.5x input.
//   Anthropic (all):          cache read = 0.1x input, 5-min cache write = 1.25x input.
//   Gemini 2.5: cache read = 0.1x input.  Gemini 1.5: cache read = 0.25x input.
export const PRICING: Record<string, ModelPricing> = {
  // OpenAI
  'gpt-4o':           { provider: 'openai', inputPerMillion: 2.5,  outputPerMillion: 10.0, cachedInputPerMillion: 1.25 },
  'gpt-4o-mini':      { provider: 'openai', inputPerMillion: 0.15, outputPerMillion: 0.6,  cachedInputPerMillion: 0.075 },
  'gpt-4-turbo':      { provider: 'openai', inputPerMillion: 10,   outputPerMillion: 30,   cachedInputPerMillion: 5 },
  'gpt-4':            { provider: 'openai', inputPerMillion: 30,   outputPerMillion: 60,   cachedInputPerMillion: 15 },
  'gpt-3.5-turbo':    { provider: 'openai', inputPerMillion: 0.5,  outputPerMillion: 1.5,  cachedInputPerMillion: 0.25 },
  'o1-preview':       { provider: 'openai', inputPerMillion: 15,   outputPerMillion: 60,   cachedInputPerMillion: 7.5 },
  'o1-mini':          { provider: 'openai', inputPerMillion: 3,    outputPerMillion: 12,   cachedInputPerMillion: 1.5 },
  // GPT-5 family (current OpenAI generation; verified against
  // developers.openai.com/api/docs/pricing on 2026-07-22, Standard tier). 5.6
  // (sol/terra/luna) is the current flagship line; 5.5 and 5.4 remain available.
  'gpt-5.6-sol':      { provider: 'openai', inputPerMillion: 5,    outputPerMillion: 30,   cachedInputPerMillion: 0.5 },
  'gpt-5.6-terra':    { provider: 'openai', inputPerMillion: 2.5,  outputPerMillion: 15,   cachedInputPerMillion: 0.25 },
  'gpt-5.6-luna':     { provider: 'openai', inputPerMillion: 1,    outputPerMillion: 6,    cachedInputPerMillion: 0.1 },
  'gpt-5.5-pro':      { provider: 'openai', inputPerMillion: 30,   outputPerMillion: 180,  cachedInputPerMillion: 3 },
  'gpt-5.5':          { provider: 'openai', inputPerMillion: 5,    outputPerMillion: 30,   cachedInputPerMillion: 0.5 },
  'gpt-5.4-pro':      { provider: 'openai', inputPerMillion: 30,   outputPerMillion: 180,  cachedInputPerMillion: 3 },
  'gpt-5.4':          { provider: 'openai', inputPerMillion: 2.5,  outputPerMillion: 15,   cachedInputPerMillion: 0.25 },
  'gpt-5.4-mini':     { provider: 'openai', inputPerMillion: 0.75, outputPerMillion: 4.5,  cachedInputPerMillion: 0.075 },
  'gpt-5.4-nano':     { provider: 'openai', inputPerMillion: 0.2,  outputPerMillion: 1.25, cachedInputPerMillion: 0.02 },
  // Anthropic. Current flagships (verified 2026-07-19): Opus 4.8/4.7 = 5/25,
  // Sonnet 4.6 = 3/15, Haiku 4.5 = 1/5. `claude-opus-4` keeps the legacy 4.0/4.1
  // rate (15/75); newer variants get explicit longer-prefix keys.
  'claude-opus-4-8':  { provider: 'anthropic', inputPerMillion: 5,   outputPerMillion: 25, cachedInputPerMillion: 0.5,  cacheWritePerMillion: 6.25 },
  'claude-opus-4-7':  { provider: 'anthropic', inputPerMillion: 5,   outputPerMillion: 25, cachedInputPerMillion: 0.5,  cacheWritePerMillion: 6.25 },
  'claude-opus-4':    { provider: 'anthropic', inputPerMillion: 15,  outputPerMillion: 75, cachedInputPerMillion: 1.5,  cacheWritePerMillion: 18.75 },
  'claude-sonnet-4-6':{ provider: 'anthropic', inputPerMillion: 3,   outputPerMillion: 15, cachedInputPerMillion: 0.3,  cacheWritePerMillion: 3.75 },
  'claude-sonnet-4':  { provider: 'anthropic', inputPerMillion: 3,   outputPerMillion: 15, cachedInputPerMillion: 0.3,  cacheWritePerMillion: 3.75 },
  'claude-haiku-4-5': { provider: 'anthropic', inputPerMillion: 1,   outputPerMillion: 5,  cachedInputPerMillion: 0.1,  cacheWritePerMillion: 1.25 },
  'claude-3-5-sonnet':{ provider: 'anthropic', inputPerMillion: 3,   outputPerMillion: 15, cachedInputPerMillion: 0.3,  cacheWritePerMillion: 3.75 },
  'claude-3-5-haiku': { provider: 'anthropic', inputPerMillion: 0.8, outputPerMillion: 4,  cachedInputPerMillion: 0.08, cacheWritePerMillion: 1.0 },
  'claude-3-opus':    { provider: 'anthropic', inputPerMillion: 15,  outputPerMillion: 75, cachedInputPerMillion: 1.5,  cacheWritePerMillion: 18.75 },
  // Gemini
  'gemini-2.5-pro':   { provider: 'gemini', inputPerMillion: 1.25,  outputPerMillion: 10, cachedInputPerMillion: 0.125 },
  'gemini-2.5-flash':      { provider: 'gemini', inputPerMillion: 0.3, outputPerMillion: 2.5, cachedInputPerMillion: 0.03 },
  'gemini-2.5-flash-lite': { provider: 'gemini', inputPerMillion: 0.1, outputPerMillion: 0.4, cachedInputPerMillion: 0.01 },
  'gemini-1.5-pro':   { provider: 'gemini', inputPerMillion: 1.25,  outputPerMillion: 5,  cachedInputPerMillion: 0.3125 },
  'gemini-1.5-flash': { provider: 'gemini', inputPerMillion: 0.075, outputPerMillion: 0.3, cachedInputPerMillion: 0.01875 },
  // Gemini 3.x (current generation; 2.x above is legacy). Verified against
  // ai.google.dev/gemini-api/docs/pricing on 2026-07-22 (text base rate).
  'gemini-3.6-flash':         { provider: 'gemini', inputPerMillion: 1.5,  outputPerMillion: 7.5,  cachedInputPerMillion: 0.15 },
  'gemini-3.5-flash':         { provider: 'gemini', inputPerMillion: 1.5,  outputPerMillion: 9.0,  cachedInputPerMillion: 0.15 },
  'gemini-3.5-flash-lite':    { provider: 'gemini', inputPerMillion: 0.3,  outputPerMillion: 2.5,  cachedInputPerMillion: 0.03 },
  'gemini-3.1-flash-lite':    { provider: 'gemini', inputPerMillion: 0.25, outputPerMillion: 1.5,  cachedInputPerMillion: 0.025 },
  'gemini-3.1-pro-preview':   { provider: 'gemini', inputPerMillion: 2.0,  outputPerMillion: 12.0, cachedInputPerMillion: 0.20 },
  'gemini-3-flash-preview':   { provider: 'gemini', inputPerMillion: 0.5,  outputPerMillion: 3.0,  cachedInputPerMillion: 0.05 },
  'gemini-flash-latest':      { provider: 'gemini', inputPerMillion: 1.5,  outputPerMillion: 7.5,  cachedInputPerMillion: 0.15 },
  'gemini-flash-lite-latest': { provider: 'gemini', inputPerMillion: 0.3,  outputPerMillion: 2.5,  cachedInputPerMillion: 0.03 },
  'gemini-pro-latest':        { provider: 'gemini', inputPerMillion: 2.0,  outputPerMillion: 12.0, cachedInputPerMillion: 0.20 },
  // Kimi (Moonshot AI). OpenAI-compatible API, so usage + cache fields match
  // OpenAI's shape. Rates are approximate and change often; verify at
  // platform.moonshot.ai before relying on them long term (lastVerified 2026-07-22):
  // K3 = 3/15, K2.6 = 0.95/4, K2.5 = 0.60/3, K2 (older tier) approx 0.60/3.
  'kimi-k3':          { provider: 'kimi', inputPerMillion: 3.0,  outputPerMillion: 15.0, cachedInputPerMillion: 0.30 },
  'kimi-k2.6':        { provider: 'kimi', inputPerMillion: 0.95, outputPerMillion: 4.0,  cachedInputPerMillion: 0.16 },
  'kimi-k2.5':        { provider: 'kimi', inputPerMillion: 0.6,  outputPerMillion: 3.0,  cachedInputPerMillion: 0.10 },
  'kimi-k2':          { provider: 'kimi', inputPerMillion: 0.6,  outputPerMillion: 3.0,  cachedInputPerMillion: 0.10 },
};

/**
 * Resolve pricing for a model name, allowing dated variants.
 *
 * Real-world IDs are date-pinned (`gpt-4o-2024-08-06`, `claude-opus-4-20250514`,
 * `claude-3-5-sonnet-latest`). Strict equality would silently log cost=0 for
 * those — which neuters every cap. So we longest-prefix-match against PRICING:
 * `claude-3-5-sonnet-20241022` resolves to `claude-3-5-sonnet`, not the
 * shorter `claude-3` family.
 */
export function resolvePricing(model: string): ModelPricing | null {
  if (PRICING[model]) return PRICING[model]!;
  const lower = model.toLowerCase();
  let best: { key: string; pricing: ModelPricing } | null = null;
  for (const [key, pricing] of Object.entries(PRICING)) {
    if (lower.startsWith(key.toLowerCase())) {
      if (!best || key.length > best.key.length) best = { key, pricing };
    }
  }
  return best?.pricing ?? null;
}

/**
 * Prompt-cache aware. `cachedInputTokens` is the cache-READ subset already
 * included in `inputTokens` (billed at the discounted rate); `cacheCreationTokens`
 * is additive cache-WRITE volume (Anthropic). Both default to 0. The server
 * recomputes and ignores this value, so it is only the SDK's local estimate.
 */
export function calculateCost(
  model: string,
  inputTokens: number,
  outputTokens: number,
  cachedInputTokens = 0,
  cacheCreationTokens = 0,
): { costUsd: number; provider: Provider | null } {
  const pricing = resolvePricing(model);
  if (!pricing) return { costUsd: 0, provider: null };
  const cached = Math.max(0, Math.min(cachedInputTokens, inputTokens));
  const uncached = inputTokens - cached;
  const cachedRate = pricing.cachedInputPerMillion ?? pricing.inputPerMillion;
  const writeRate = pricing.cacheWritePerMillion ?? pricing.inputPerMillion;
  const cost =
    (uncached / 1_000_000) * pricing.inputPerMillion +
    (cached / 1_000_000) * cachedRate +
    (Math.max(0, cacheCreationTokens) / 1_000_000) * writeRate +
    (outputTokens / 1_000_000) * pricing.outputPerMillion;
  return { costUsd: round6(cost), provider: pricing.provider };
}

function round6(n: number): number {
  return Math.round(n * 1_000_000) / 1_000_000;
}

File: python/weckr/pricing.py
from __future__ import annotations

"""Per-million-token pricing + cheaper-alternative mapping.

``calculate_cost(model, input_tokens, output_tokens)`` returns USD cost as a
float. Unknown models return 0.0 — server-side recalculation in
/api/v1/log uses the same prefix-match so unknown models can be priced too.

KEEP IN SYNC with:
  - src/pricing.ts             (TypeScript SDK)
  - weckr-api/lib/caps.ts      (server cost recalc)
If you add a model here, add it there too.
"""

from typing import Dict, TypedDict


class ModelPricing(TypedDict):
    input: float
    output: float
    # cache-READ rate per million (discounted repeated context).
    cached_input: float
    # cache-WRITE rate per million. Anthropic charges a premium; for OpenAI/Gemini
    # (no per-request write charge) this equals `input` and is never used, since
    # those providers never report cache-creation tokens.
    cache_write: float


# Per-million-token pricing for supported models.
#
# Cached rates verified against official provider pricing on 2026-07-18:
#   OpenAI gpt-4o + o-series: cache read = 0.5x input.
#   Anthropic (all):          cache read = 0.1x input, 5-min cache write = 1.25x input.
#   Gemini 2.5: cache read = 0.1x input.  Gemini 1.5: cache read = 0.25x input.
PRICING: Dict[str, ModelPricing] = {
    # OpenAI
    "gpt-4o":           {"input": 2.50,  "output": 10.00, "cached_input": 1.25,  "cache_write": 2.50},
    "gpt-4o-mini":      {"input": 0.15,  "output": 0.60,  "cached_input": 0.075, "cache_write": 0.15},
    "gpt-4-turbo":      {"input": 10.00, "output": 30.00, "cached_input": 5.00,  "cache_write": 10.00},
    "gpt-4":            {"input": 30.00, "output": 60.00, "cached_input": 15.00, "cache_write": 30.00},
    "gpt-3.5-turbo":    {"input": 0.50,  "output": 1.50,  "cached_input": 0.25,  "cache_write": 0.50},
    "o1-preview":       {"input": 15.00, "output": 60.00, "cached_input": 7.50,  "cache_write": 15.00},
    "o1-mini":          {"input": 3.00,  "output": 12.00, "cached_input": 1.50,  "cache_write": 3.00},
    # Anthropic. Current flagships (verified 2026-07-19): Opus 4.8/4.7 = 5/25,
    # Sonnet 4.6 = 3/15, Haiku 4.5 = 1/5. "claude-opus-4" keeps the legacy
    # 4.0/4.1 rate (15/75); newer variants get explicit longer-prefix keys.
    "claude-opus-4-8":   {"input": 5.00,  "output": 25.00, "cached_input": 0.50,  "cache_write": 6.25},
    "claude-opus-4-7":   {"input": 5.00,  "output": 25.00, "cached_input": 0.50,  "cache_write": 6.25},
    "claude-opus-4":     {"input": 15.00, "output": 75.00, "cached_input": 1.50,  "cache_write": 18.75},
    "claude-sonnet-4-6": {"input": 3.00,  "output": 15.00, "cached_input": 0.30,  "cache_write": 3.75},
    "claude-sonnet-4":   {"input": 3.00,  "output": 15.00, "cached_input": 0.30,  "cache_write": 3.75},
    "claude-haiku-4-5":  {"input": 1.00,  "output": 5.00,  "cached_input": 0.10,  "cache_write": 1.25},
    "claude-3-5-sonnet": {"input": 3.00, "output": 15.00, "cached_input": 0.30,  "cache_write": 3.75},
    "claude-3-5-haiku": {"input": 0.80,  "output": 4.00,  "cached_input": 0.08,  "cache_write": 1.00},
    "claude-3-opus":    {"input": 15.00, "output": 75.00, "cached_input": 1.50,  "cache_write": 18.75},
    # Gemini
    "gemini-2.5-pro":   {"input": 1.25,  "output": 10.00, "cached_input": 0.125,   "cache_write": 1.25},
    "gemini-2.5-flash": {"input": 0.15,  "output": 0.60,  "cached_input": 0.015,   "cache_write": 0.15},
    "gemini-1.5-pro":   {"input": 1.25,  "output": 5.00,  "cached_input": 0.3125,  "cache_write": 1.25},
    "gemini-1.5-flash": {"input": 0.075, "output": 0.30,  "cached_input": 0.01875, "cache_write": 0.075},
}


# Cheaper alternative per model — used when a cap's action is "downgrade".
# Same-provider only; never silently switch a customer to a different vendor.
CHEAPER_ALTERNATIVE: Dict[str, str] = {
    # OpenAI
    "gpt-4o":          "gpt-4o-mini",
    "gpt-4-turbo":     "gpt-4o-mini",
    "gpt-4":           "gpt-4o-mini",
    # Anthropic
    "claude-opus-4-8":   "claude-sonnet-4-6",
    "claude-opus-4-7":   "claude-sonnet-4-6",
    "claude-opus-4":     "claude-sonnet-4",
    "claude-sonnet-4-6": "claude-haiku-4-5",
    "claude-sonnet-4":   "claude-haiku-4-5",
    # Gemini
    "gemini-2.5-pro":  "gemini-2.5-flash",
    "gemini-1.5-pro":  "gemini-2.5-flash",
}


def resolve_pricing(model: str) -> ModelPricing:
    """Resolve pricing for a model name, allowing dated variants.

    Real-world IDs are date-pinned (``gpt-4o-2024-08-06``,
    ``claude-3-5-sonnet-latest``). Strict equality would silently log cost=0
    for those — which neuters every cap. So we longest-prefix-match against
    PRICING: ``claude-3-5-sonnet-20241022`` resolves to ``claude-3-5-sonnet``,
    not the shorter ``claude-3`` family.

    Returns an empty pricing dict ``{"input": 0, "output": 0}`` if no match.
    """
    if model in PRICING:
        return PRICING[model]
    lower = model.lower()
    best_key: str = ""
    best_pricing: ModelPricing = {
        "input": 0.0,
        "output": 0.0,
        "cached_input": 0.0,
        "cache_write": 0.0,
    }
    for key, pricing in PRICING.items():
        if lower.startswith(key.lower()):
            if len(key) > len(best_key):
                best_key = key
                best_pricing = pricing
    return best_pricing


def calculate_cost(
    model: str,
    input_tokens: int,
    output_tokens: int,
    cached_input_tokens: int = 0,
    cache_creation_tokens: int = 0,
) -> float:
    """Return USD cost for ``model`` given token counts.

    Prompt-cache aware. ``cached_input_tokens`` is the cache-READ subset already
    included in ``input_tokens`` (billed at the discounted cached rate);
    ``cache_creation_tokens`` is additive cache-WRITE volume (Anthropic). Both
    default to 0, so callers that don't pass them price exactly as before.

    Unknown models return 0.0 (best-effort; server-side recompute will catch).
    """
    pricing = resolve_pricing(model)
    cached = max(0, min(cached_input_tokens, input_tokens))
    uncached = input_tokens - cached
    cached_rate = pricing.get("cached_input", pricing["input"])
    write_rate = pricing.get("cache_write", pricing["input"])
    return (
        uncached * pricing["input"]
        + cached * cached_rate
        + max(0, cache_creation_tokens) * write_rate
        + output_tokens * pricing["output"]
    ) / 1_000_000.0


__all__ = [
    "ModelPricing",
    "PRICING",
    "CHEAPER_ALTERNATIVE",
    "resolve_pricing",
    "calculate_cost",
]
